projects/file-parsing/main.go

```diff
@@ -0,0 +1,360 @@
+		return errors.New("[3] could not parse any data in CSV format")
+	}
+
+	// loop over each line
+	for _, line := range lines {
+
+		// (!!!) if we try to read one of the binary files it gives us this: [[��] [AyaPrisha����CharlieMargot]]
+		// so we need to add an additional check here:
+
+		// if the line does not have 2 values (highscore, name) then skip
+		// ((!) this is brittle if a line has valid data but random extra values etc)
+		if len(line) != 2 {
+			continue
+		}
+
+		// get the player name
+		name := line[0]
+
+		// the player has no name so skip
+		if len(name) < 1 {
+			continue
+		}
+
+		//  parse the string to get the score ((!) this is of type int not int32)
+		highScore, err := strconv.Atoi(line[1])
+
+		// if the player has no high score then skip
+		if err != nil {
+			continue
+		}
+
+		// add the player to the dataSlice
+		*dataSlice = append(*dataSlice, Player{Name: name, HighScore: int32(highScore)})
+	}
+
+	// if the data slice is empty
+	if len(*dataSlice) == 0 {
+		return errors.New("[3] could not parse any data in CSV format")
+	}
+
+	return nil
+}
+
+// [4] BINARY
+func parseBinary(file []byte, dataSlice *[]Player) error {
+	// each record contains exactly four bytes
+	// representing the score as a signed 32-bit integer (in the above endian format)
+	// then the name of the player stored in UTF-8
+	// which may not contain a null character, followed by a null terminating character
+
+	// --- DEBUG PRINT TABLE
+	// for index := 0; index < len(file); index++ {
+	// 	// print out the bytes in the slice in various formats
+	// 	fmt.Printf("slice index: %2d | Binary: %08b | Bytes (as an Integer): %3v | Hexadecimal: %2x | Decimal: %3d | ASCII : %c\n", index, file[index], file[index], file[index], file[index], file[index])
+	// }
+
+	// BIG ENDIAN (starts with fe ff) (00 at the end is the null terminator)
+	// data is displayed in groups of 2 bytes
+	// each group representing a hexadecimal value
+	// byte order where most significant byte is stored first
+
+	// 0000000 [fe ff] [00 00 00 0a] 41 79 61 00 00 00 00 1e 50 72
+	// 0000020 69 73 68 61 00 ff ff ff ff 43 68 61 72 6c 69 65
+	// 0000040 00 00 00 00 19 4d 61 72 67 6f 74 [00]
+	// 0000054
+
+	// LITTLE ENDIAN (starts with ff fe) (00 at the end is the null terminator)
+	// data is displayed in groups of 2 bytes
+	// each group representing a hexadecimal value
+	// byte order where least significant byte is stored first
+
+	// 0000000 [ff fe] [0a 00 00 00] 41 79 61 00 1e 00 00 00 50 72
+	// 0000020 69 73 68 61 00 ff ff ff ff 43 68 61 72 6c 69 65
+	// 0000040 00 19 00 00 00 4d 61 72 67 6f 74 [00]
+	// 0000054
+
+	// initialise two flags (explicitly two, yes could be achieved with one)
+	var bigEndian bool
+	var littleEndian bool
+
+	// check the first two bytes of the file (deliberately using byte() to be explicit)
+	// if they are "fe ff" it is big endian
+	// if they are "ff fe" is is little endian
+	if file[0] == byte(0xfe) && file[1] == byte(0xff) {
```

Comment:
I believe this line has a couple of ways it could panic - can you spot them?

Reply:
the one i can immediately think of:
- if the file passed into the function is empty it will attempt to access a byte at index 0 and that won't be found and will panic